### DISCUSSION TITLE
Phoebus and CS-Studio databrowser comparison

### DISCUSSION THREAD

**Discussion by willrogers** [ControlSystemStudio/phoebus#1796] [General]
Looking at recent changes to the databrowser in CS-Studio and the latest version in Phoebus.

**Comment by rjwills28** [comment]
I have identified a few useful changes that we have recently made to CS-Studio that might be worth also integrating with Phoebus. 

- We requested a change to the AA to include a new post-processor called 'Optimized With Last Sample'. This is based on the 'Optimized' post-processor but if a bin contains no events then instead of simply copying the previous bin, it uses the last sample from the last bin to actually contain events. We have added an option in the databrowser preferences to use this new post-processor and in testing we have found it to provide a more accurate representation of the raw data. Further details can be found here: https://github.com/ControlSystemStudio/cs-studio/pull/2675.
- Remove the standard deviation lines from the databrowser plot when plotting an AREA trace (see discussion https://github.com/ControlSystemStudio/phoebus/issues/1656)
- Display enum labels in the databrowser 'Cursor value' field and when using the Cross-hair tool. For example, instead of just displaying (1), the label would show 'Open (1)'. Further details here: https://github.com/ControlSystemStudio/cs-studio/pull/2662.
- We have improved the 'annotation' usability by making it possible to edit annotations simply by 'double-clicking' on them and also being able to right-click in the databrowser and select 'Add annotation' from the context menu. Finally we allowed the user to define the format for the value and data, i.e. they can specify how many dp the value should be written to. Details here: https://github.com/ControlSystemStudio/cs-studio/pull/2674.
- We found a bug when getting data back from the Archiver Appliance using the 'Optimized' post-processor. If the PV has not changed within the timeframe requested, then the AA returns an Iterator but it is empty. This does not get checked by the databrowser and results in no data being plotted in the databrowser. More details here: https://github.com/ControlSystemStudio/cs-studio/pull/2665.

Any thoughts/opinions as to which features should be implemented in Phoebus?

**Reply by kasemir** [reply]
Makes sense to me to implement all of these in the latest version as well.

**Reply by willrogers** [reply]
Here is an example of updated CS-Studio vs Phoebus databrowser using the AA. This is the DLS beam current. At no point did it go above 310 mA or below 0 mA. I think it is plotting the standard deviation that has this misleading effect.

![Screenshot (1)](https://user-images.githubusercontent.com/2440055/110611632-4d288000-8187-11eb-9941-558dc9946322.png)

**Comment by rjwills28** [comment]
Another thing I have noticed that occurs in both CS-Studio & Phoebus is that if you choose to not show a trace (i.e. deselect the 'Show' option in the 'Traces' panel), the trace disappears but any annotations that have been added do not. They remain visible. Looks a bit confusing if you start showing other traces.

**Reply by rjwills28** [reply]
I have been having a look into this issue. Presumably the desired behaviour would be to have the annotation disappear with the trace and then re-appear when the trace is 'shown' again? The simplest solution I have removes the annotation with the trace but does not bring it back when the trace is drawn again so I need to do a bit more work on that. Do we think this is the correct approach? @kasemir

**Reply by kasemir** [reply]
I would not remove/add the annotation but simply skip the drawing.

Each `AnnotationImpl` has a reference to the associated `trace`.
I would simply use that inside `paint`:
```
/** Paint the annotation on given gc and axes. */
void paint(final Graphics2D gc, final AxisPart<XTYPE> xaxis, final YAxisImpl<XTYPE> yaxis)
{
    // Skip drawing annotation of hidden traces
    if (! trace.isVisible())
        return;
   // .. otherwise continue with drawing the annotation... 
```

Similarly, you might need to check `trace.isVisible()` in the annotation code that detects mouse clicks, so it's not reacting to them when the trace is hidden.

**Reply by rjwills28** [reply]
Thanks for the input on this. The only problem I have found trying to implement this is that the actual trace is removed from plot when it is deselected. Then when it is selected again it is added. 

So when the trace is removed and `trace.isVisible() `= false, then the annotation is hidden fine. But when the trace is added again it is done as a new trace and so the annotation has no link to that trace (`trace.isVisible()` continues to be false) and so that annotation does not appear in the databrowser.

**Reply by kasemir** [reply]
I think you're correct. The `AnnotationImpl` won't have a `trace` any more because the trace has been removed from the plot..
But I think the trace is still in the `Model`. The `Model` has all `ModelItem`s and all `AnnotationInfo`s. When a model item is hidden, the corresponding trace is removed from the UI. And I think you're correct in that the annotation also needs to be removed.
When later re-enabling the model item, it's added back to the plot.  .. and it should be possible to add the annotation back to the plot as well because the information is still in the model.

**Comment by rjwills28** [comment]
Slightly unrelated to the databrowser but we have also identified an issue with the slider widget when the slider limits are set with 'limits from PV'. If this is the case then it tries to get the PV control limits (i.e. values of DRVL and DRVH) and if these are not set then it uses the display limits (i.e. LOPR and HOPR). If both of these are not set then it uses:
- In Phoebus: hardcoded value of 0 and 100.  
- In CS-Studio: limits of 0,0 end up setting slider limits to 0 and 1. 

Certainly the CS-Studio is probably not correct but I also wonder about using the hardcoded limits if 0 and 100. Could we not instead fall back to the min/max that the user can define in the widget configuration (i.e. in Phoebus I think this would be something like `model_widget.propMinimum().getValue()` ) ?

**Reply by kasemir** [reply]
I guess there's no 'correct' option. If you use 'limits from PV', then the PV should have limits.
Still, I like the idea of falling back to the minimum/maximum properties instead of hardcoded 0..100.
